src/sentry/notifications/notificationcontroller.py

```diff
@@ -0,0 +1,460 @@
+            for item in scope_items:
+                scope: Scope = (scope_type, item)
+                for type in NotificationSettingEnum:
+                    most_specific_setting = None
+                    for setting in scoped_settings:
+                        user_settings = self._filter_options(
+                            settings=setting, user_id=recipient.id, type=type.value
+                        )
+                        team_settings = self._filter_options(
+                            settings=setting, team_id=recipient.id, type=type.value
+                        )
+                        s = user_settings + team_settings
+                        if len(s) > 0:
+                            most_specific_setting = NotificationSettingsOptionEnum(s[0].value)
+                            break
+
+                    if most_specific_setting is None and type in defaults:
+                        most_specific_setting = defaults[type]
+
+                    if most_specific_setting is None:
+                        continue
+
+                    layered_setting_options[recipient][scope][type] = most_specific_setting
+
+        return layered_setting_options
+
+    def _get_layered_setting_providers(
+        self,
+        **kwargs,
+    ) -> MutableMapping[
+        Recipient,
+        MutableMapping[
+            Scope,
+            MutableMapping[
+                NotificationSettingEnum,
+                MutableMapping[ExternalProviderEnum, NotificationSettingsOptionEnum],
+            ],
+        ],
+    ]:
+        """
+        Returns a mapping of the most specific notification setting providers for the given recipients and scopes.
+        Note that this includes default settings for any notification types that are not set.
+        """
+
+        settings = self.get_all_setting_providers()
+        project_settings = self._filter_providers(
+            settings=settings, scope_type=NotificationScopeEnum.PROJECT.value, **kwargs
+        )
+        org_settings = self._filter_providers(
+            settings=settings, scope_type=NotificationScopeEnum.ORGANIZATION.value, **kwargs
+        )
+        user_settings = self._filter_providers(
+            settings=settings, scope_type=NotificationScopeEnum.USER.value, **kwargs
+        )
+        scoped_settings = [project_settings, org_settings, user_settings]
+
+        defaults = get_provider_defaults()
+        layered_setting_providers: MutableMapping[
+            Recipient,
+            MutableMapping[
+                Scope,
+                MutableMapping[
+                    NotificationSettingEnum,
+                    MutableMapping[ExternalProviderEnum, NotificationSettingsOptionEnum],
+                ],
+            ],
+        ] = defaultdict(lambda: defaultdict(lambda: defaultdict(dict)))
+        for recipient in self.recipients:
+            (scope_type, scope_items) = self._get_scope_for_layering(recipient)
+            for item in scope_items:
```

Comment:
@snigdhas what happens if there are no scope items? Seems like the output would be an empty dict which doesn't handle defaults correctly. Email is on by default so if the user has no settings at all, we should still expect email to be enabled.

Reply:
IIUC there should always be scope items or else we raise an exception

```python
if self.project_ids:
    return (NotificationScopeEnum.PROJECT, self.project_ids)
if self.organization_id is not None:
    return (NotificationScopeEnum.ORGANIZATION, [self.organization_id])
if recipient_is_user(recipient):
    return (NotificationScopeEnum.USER, [recipient])
if recipient_is_team(recipient):
    return (NotificationScopeEnum.TEAM, [recipient])

raise ValueError(f"Recipient is of unsupported type: {recipient}")
```

Reply:
@snigdhas oh yea good call